cirq/work/pauli_string_collector.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+from typing import Optional, MutableMapping, cast, Union
+
+import numpy as np
+
+from cirq import circuits, study, ops, value
+from cirq.work import collector
+
+
+class PauliStringSampleCollector(collector.SampleCollector):
+    """Estimates the energy of a linear combination of Pauli observables."""
+
+    def __init__(self,
+                 circuit: circuits.Circuit,
+                 samples_per_term: int,
+                 terms: value.LinearDict[ops.PauliString],
+                 max_samples_per_job: int = 1000000):
+        """
+        Args:
+            circuit: Produces the state to be tested.
+            terms: The pauli product observables to measure. Their sampled
+                expectations will be scaled by their coefficients and their
+                dictionary weights, and then added up to produce the final
+                result.
+            max_samples_per_job: How many samples to request at a time.
+        """
+        self._circuit = circuit
+        self._samples_per_job = max_samples_per_job
+        self._terms = [
+            # Merge coefficients.
+            (pauli_string / pauli_string.coefficient,
+             coef * pauli_string.coefficient)
+            for pauli_string, coef in terms.items()
+        ]
+        self._zeros = collections.defaultdict(
+            lambda: 0)  # type: MutableMapping[ops.PauliString, int]
+        self._ones = collections.defaultdict(
+            lambda: 0)  # type: MutableMapping[ops.PauliString, int]
+        self._samples_per_term = samples_per_term
+        self._total_samples_requested = 0
+
+    def next_job(self) -> Optional[collector.CircuitSampleJob]:
```

Comment:
this is sort of explicitly the pattern we don't want to follow for this type of estimation.  It causes the jobs to be broken up into pieces that could be combined using parameterizations.